StaticRoute cannot infer type for optional static_params and panics when provided a default

**Describe the bug**
Attempting to use the new `<StaticRoute />` feature in 0.5.0

**Leptos Dependencies**

Please copy and paste the Leptos dependencies and features from your `Cargo.toml`.

For example:
```toml
leptos = { version = "0.5", features = ["rustls"] }
leptos_meta = { version = "0.5" }
leptos_dom = { version = "0.5" }
leptos_router = { version = "0.5" }
leptos_axum = { version = "0.5", optional = true }
```

**To Reproduce**

1. Create a new,  minimal leptos project with the above dependencies (e.g. `todo_app_sqlite_axum`)
2. Replace the existing routes line with the static generation lines in `main.rs`:
```rust
let (routes, static_data_map) = leptos_axum::generate_route_list_with_ssg(TodoApp);
leptos_router::build_static_routes(&leptos_options, TodoApp, &routes, &static_data_map).await.unwrap();
```
4. Change the `<Route />` component in `todo.rs` to `<StaticRoute />`:
```rust
<StaticRoute path="" view=Todos />
```
5. Get a type inference error:
```txt
error[E0282]: type annotations needed
   --> src/todo.rs:108:22
    |
108 |                     <StaticRoute path="" view=Todos/>
    |                      ^^^^^^^^^^^ cannot infer type of the type parameter `S` declared on the function `StaticRoute`
    |
help: consider specifying the generic arguments
    |
108 |                     <StaticRoute::<_, _, &str, S> path="" view=Todos/>
    |                                 +++++++++++++++++
```
7. Add the `static_params` prop:
```rust
<StaticRoute path="" view=Todos static_params=|| Box::pin(async { StaticParamsMap::default() }) />
```
9. Run `cargo leptos watch`
10. Receive a panic:
```txt
building static route:
thread 'main' panicked at /home/luke/dev/leptos/leptos_reactive/src/spawn.rs:87:13:
`spawn_local` called from outside of a `task::LocalSet`
stack backtrace:
   0: rust_begin_unwind
             at /rustc/187b8131d4f760f856b214fce34534903276f2ef/library/std/src/panicking.rs:597:5
   1: core::panicking::panic_fmt
             at /rustc/187b8131d4f760f856b214fce34534903276f2ef/library/core/src/panicking.rs:72:14
   2: tokio::task::local::spawn_local_inner
             at /home/luke/.cargo/registry/src/index.crates.io-6f17d22bba15001f/tokio-1.32.0/src/task/local.rs:347:21
   3: tokio::task::local::spawn_local
             at /home/luke/.cargo/registry/src/index.crates.io-6f17d22bba15001f/tokio-1.32.0/src/task/local.rs:337:9
   4: leptos_reactive::spawn::spawn_local
             at /home/luke/dev/leptos/leptos_reactive/src/spawn.rs:87:13
   5: leptos_reactive::resource::ResourceState<S,T>::load::{{closure}}
             at /home/luke/dev/leptos/leptos_reactive/src/resource.rs:1337:13
   6: leptos_reactive::memo::forward_ref_to::{{closure}}
             at /home/luke/dev/leptos/leptos_reactive/src/memo.rs:258:9
   7: leptos_reactive::signal::<impl leptos_reactive::node::NodeId>::try_with_no_subscription
             at /home/luke/dev/leptos/leptos_reactive/src/signal.rs:2041:12
   8: <leptos_reactive::memo::Memo<T> as leptos_reactive::signal::SignalWith>::try_with::{{closure}}
             at /home/luke/dev/leptos/leptos_reactive/src/memo.rs:472:13
   9: leptos_reactive::runtime::with_runtime::{{closure}}
             at /home/luke/dev/leptos/leptos_reactive/src/runtime.rs:700:41
  10: std::thread::local::LocalKey<T>::try_with
             at /rustc/187b8131d4f760f856b214fce34534903276f2ef/library/std/src/thread/local.rs:270:16
  11: std::thread::local::LocalKey<T>::with
             at /rustc/187b8131d4f760f856b214fce34534903276f2ef/library/std/src/thread/local.rs:246:9
  12: leptos_reactive::runtime::with_runtime
             at /home/luke/dev/leptos/leptos_reactive/src/runtime.rs:695:13
  13: <leptos_reactive::memo::Memo<T> as leptos_reactive::signal::SignalWith>::try_with
             at /home/luke/dev/leptos/leptos_reactive/src/memo.rs:470:9
  14: leptos_reactive::resource::ResourceState<S,T>::load
             at /home/luke/dev/leptos/leptos_reactive/src/resource.rs:1309:13
  15: leptos_reactive::resource::load_resource::{{closure}}
             at /home/luke/dev/leptos/leptos_reactive/src/resource.rs:391:13
  16: std::thread::local::LocalKey<T>::try_with
             at /rustc/187b8131d4f760f856b214fce34534903276f2ef/library/std/src/thread/local.rs:270:16
  17: std::thread::local::LocalKey<T>::with
             at /rustc/187b8131d4f760f856b214fce34534903276f2ef/library/std/src/thread/local.rs:246:9
  18: leptos_reactive::resource::load_resource
             at /home/luke/dev/leptos/leptos_reactive/src/resource.rs:389:5
  19: leptos_reactive::resource::create_resource_helper::{{closure}}
             at /home/luke/dev/leptos/leptos_reactive/src/resource.rs:230:13
  20: <leptos_reactive::effect::EffectState<T,F> as leptos_reactive::effect::AnyComputation>::run
             at /home/luke/dev/leptos/leptos_reactive/src/effect.rs:366:25
  21: leptos_reactive::runtime::Runtime::update::{{closure}}
             at /home/luke/dev/leptos/leptos_reactive/src/runtime.rs:230:25
  22: leptos_reactive::runtime::Runtime::with_observer
             at /home/luke/dev/leptos/leptos_reactive/src/runtime.rs:328:17
  23: leptos_reactive::runtime::Runtime::update
             at /home/luke/dev/leptos/leptos_reactive/src/runtime.rs:226:21
  24: leptos_reactive::runtime::Runtime::update_if_necessary
             at /home/luke/dev/leptos/leptos_reactive/src/runtime.rs:186:13
  25: leptos_reactive::effect::create_isomorphic_effect::{{closure}}
             at /home/luke/dev/leptos/leptos_reactive/src/effect.rs:256:9
  26: leptos_reactive::runtime::with_runtime::{{closure}}
             at /home/luke/dev/leptos/leptos_reactive/src/runtime.rs:700:41
  27: std::thread::local::LocalKey<T>::try_with
             at /rustc/187b8131d4f760f856b214fce34534903276f2ef/library/std/src/thread/local.rs:270:16
  28: std::thread::local::LocalKey<T>::with
             at /rustc/187b8131d4f760f856b214fce34534903276f2ef/library/std/src/thread/local.rs:246:9
  29: leptos_reactive::runtime::with_runtime
             at /home/luke/dev/leptos/leptos_reactive/src/runtime.rs:695:13
  30: leptos_reactive::effect::create_isomorphic_effect
             at /home/luke/dev/leptos/leptos_reactive/src/effect.rs:255:9
  31: leptos_reactive::resource::create_resource_helper
             at /home/luke/dev/leptos/leptos_reactive/src/resource.rs:227:5
  32: leptos_reactive::resource::create_resource_with_initial_value
             at /home/luke/dev/leptos/leptos_reactive/src/resource.rs:125:5
  33: leptos_reactive::resource::create_resource
             at /home/luke/dev/leptos/leptos_reactive/src/resource.rs:92:5
  34: todo_app_sqlite_axum::todo::Todos::__Todos
             at ./src/todo.rs:122:17
  35: todo_app_sqlite_axum::todo::Todos::{{closure}}
             at ./src/todo.rs:115:1
```

**Expected behavior**
I'd expect an optional `static_params` prop to actually be optional, and to not panic when provided.

Reference issues: #1594 and PR: #1649

--- a/integrations/axum/src/lib.rs
+++ b/integrations/axum/src/lib.rs
@@ -1303,6 +1303,7 @@ pub async fn build_static_routes<IV>(
             &static_data_map,
         )
         .await
+        .expect("could not build static routes")
     });
 }
 

--- a/integrations/axum/src/lib.rs
+++ b/integrations/axum/src/lib.rs
@@ -1070,38 +1070,41 @@ where
                 let full_path = format!("http://leptos.dev{path}");
 
                 let (tx, rx) = futures::channel::oneshot::channel();
-                let local_pool = get_leptos_pool();
-                local_pool.spawn_pinned(move || {
-                    async move {
-                        let app = {
-                            let full_path = full_path.clone();
-                            let (req, req_parts) = generate_request_and_parts(req).await;
-                            move || {
-                                provide_contexts(full_path, req_parts, req.into(), default_res_options);
-                                app_fn().into_view()
-                            }
-                        };
-
-                        let (stream, runtime) =
-                            render_to_stream_in_order_with_prefix_undisposed_with_context(
-                                app,
-                                || "".into(),
-                                add_context,
+                spawn_task!(async move {
+                    let app = {
+                        let full_path = full_path.clone();
+                        let (req, req_parts) =
+                            generate_request_and_parts(req).await;
+                        move || {
+                            provide_contexts(
+                                full_path,
+                                req_parts,
+                                req.into(),
+                                default_res_options,
                             );
+                            app_fn().into_view()
+                        }
+                    };
 
-                        // Extract the value of ResponseOptions from here
-                        let res_options =
-                            use_context::<ResponseOptions>().unwrap();
+                    let (stream, runtime) =
+                        render_to_stream_in_order_with_prefix_undisposed_with_context(
+                            app,
+                            || "".into(),
+                            add_context,
+                        );
 
-                        let html = build_async_response(stream, &options, runtime).await;
+                    // Extract the value of ResponseOptions from here
+                    let res_options = use_context::<ResponseOptions>().unwrap();
 
-                        let new_res_parts = res_options.0.read().clone();
+                    let html =
+                        build_async_response(stream, &options, runtime).await;
 
-                        let mut writable = res_options2.0.write();
-                        *writable = new_res_parts;
+                    let new_res_parts = res_options.0.read().clone();
 
-                        _ = tx.send(html);
-                    }
+                    let mut writable = res_options2.0.write();
+                    *writable = new_res_parts;
+
+                    _ = tx.send(html);
                 });
 
                 let html = rx.await.expect("to complete HTML rendering");
@@ -1279,6 +1282,28 @@ where
     IV: IntoView + 'static,
 {
     generate_route_list_with_exclusions_and_ssg(app_fn, excluded_routes).0
+}
+
+/// TODO docs
+pub async fn build_static_routes<IV>(
+    options: &LeptosOptions,
+    app_fn: impl Fn() -> IV + 'static + Send + Clone,
+    routes: &[RouteListing],
+    static_data_map: StaticDataMap,
+) where
+    IV: IntoView + 'static,
+{
+    let options = options.clone();
+    let routes = routes.to_owned();
+    spawn_task!(async move {
+        leptos_router::build_static_routes(
+            &options,
+            app_fn,
+            &routes,
+            &static_data_map,
+        )
+        .await
+    });
 }
 
 /// Generates a list of all routes defined in Leptos's Router in your app. We can then use this to automatically
@@ -1493,8 +1518,7 @@ where
 
                     async move {
                         let (tx, rx) = futures::channel::oneshot::channel();
-                        let local_pool = get_leptos_pool();
-                        local_pool.spawn_pinned(move || async move {
+                        spawn_task!(async move {
                             let res = incremental_static_route(
                                 tokio::fs::read_to_string(static_file_path(
                                     &options, &path,
@@ -1537,8 +1561,7 @@ where
 
                     async move {
                         let (tx, rx) = futures::channel::oneshot::channel();
-                        let local_pool = get_leptos_pool();
-                        local_pool.spawn_pinned(move || async move {
+                        spawn_task!(async move {
                             let res = upfront_static_route(
                                 tokio::fs::read_to_string(static_file_path(
                                     &options, &path,

--- a/router/src/components/route.rs
+++ b/router/src/components/route.rs
@@ -10,6 +10,7 @@ use std::{
     future::Future,
     pin::Pin,
     rc::Rc,
+    sync::Arc,
 };
 
 thread_local! {
@@ -179,7 +180,10 @@ where
     E: IntoView,
     F: Fn() -> E + 'static,
     P: std::fmt::Display,
-    S: Fn() -> Pin<Box<dyn Future<Output = StaticParamsMap>>> + 'static,
+    S: Fn() -> Pin<Box<dyn Future<Output = StaticParamsMap> + Send + Sync>>
+        + Send
+        + Sync
+        + 'static,
 {
     define_route(
         children,
@@ -189,7 +193,7 @@ where
         &[Method::Get],
         data,
         Some(mode),
-        static_params.map(|s| Rc::new(s) as _),
+        static_params.map(|s| Arc::new(s) as _),
     )
 }
 

--- a/router/src/components/route.rs
+++ b/router/src/components/route.rs
@@ -163,8 +163,7 @@ pub fn StaticRoute<E, F, P, S>(
     /// or `|| view! { <MyComponent/>` } or even, for a component with no props, `MyComponent`).
     view: F,
     /// Creates a map of the params that should be built for a particular route.
-    #[prop(optional)]
-    static_params: Option<S>,
+    static_params: S,
     /// The static route mode
     #[prop(optional)]
     mode: StaticMode,
@@ -193,7 +192,7 @@ where
         &[Method::Get],
         data,
         Some(mode),
-        static_params.map(|s| Arc::new(s) as _),
+        Some(Arc::new(static_params)),
     )
 }
 

--- a/router/src/components/static_render.rs
+++ b/router/src/components/static_render.rs
@@ -15,7 +15,7 @@ use std::{
     hash::{Hash, Hasher},
     path::PathBuf,
     pin::Pin,
-    rc::Rc,
+    sync::Arc,
 };
 
 #[derive(Debug, Default, Serialize, Deserialize)]
@@ -304,10 +304,12 @@ pub fn purge_all_static_routes<IV>(
     purge_dir_of_static_files(Path::new(&options.site_root).to_path_buf())
 }
 
-pub type StaticData = Rc<StaticDataFn>;
+pub type StaticData = Arc<StaticDataFn>;
 
-pub type StaticDataFn =
-    dyn Fn() -> Pin<Box<dyn Future<Output = StaticParamsMap>>> + 'static;
+pub type StaticDataFn = dyn Fn() -> Pin<Box<dyn Future<Output = StaticParamsMap> + Send + Sync>>
+    + Send
+    + Sync
+    + 'static;
 
 pub type StaticDataMap = HashMap<String, Option<StaticData>>;
 

